diff --git a/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp b/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp
index 9df7d682cd3c..9471aaf57d76 100644
--- a/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp
+++ b/mlir/lib/Dialect/Linalg/IR/LinalgOps.cpp
@@ -2430,6 +2430,21 @@ public:
   }
 };
 
+struct FoldBroadcastFill : OpRewritePattern<BroadcastOp> {
+  using OpRewritePattern<BroadcastOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(BroadcastOp broadcastOp,
+                                PatternRewriter &rewriter) const override {
+    auto fillOp = broadcastOp.getInput().getDefiningOp<linalg::FillOp>();
+    if (!fillOp)
+      return failure();
+
+    rewriter.replaceOpWithNewOp<linalg::FillOp>(broadcastOp, fillOp.getInputs(),
+                                                broadcastOp.getInit());
+
+    return success();
+  }
+};
+
 } // namespace
 #endif
 
@@ -2437,7 +2452,8 @@ void BroadcastOp::getCanonicalizationPatterns(RewritePatternSet &results,
                                               MLIRContext *context) {
   results.add<EraseIdentityLinalgOp<BroadcastOp>>(context);
 #if BSPUB_DAVINCI_BISHENGIR
-  results.add<NormalizeBroadcastDenseSplatToFillConstant>(context);
+  results.add<NormalizeBroadcastDenseSplatToFillConstant, FoldBroadcastFill>(
+      context);
 #endif
 }
 
