diff --git a/mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp b/mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp
index c209e775706d..70d35a080d32 100644
--- a/mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp
+++ b/mlir/lib/Dialect/MemRef/IR/MemRefOps.cpp
@@ -2245,7 +2245,7 @@ struct ReinterpretCastReturnTypeCanonicalizer {
     dispatchIndexOpFoldResults(mixedSizes, dynamicSizes, staticSizes);
     dispatchIndexOpFoldResults(mixedStrides, dynamicStrides, staticStrides);
 
-    auto sourceMemRefType = cast<MemRefType>(op.getSource().getType());
+    auto sourceMemRefType = cast<BaseMemRefType>(op.getSource().getType());
     auto resType =
         MemRefType::get(staticSizes, sourceMemRefType.getElementType(),
                         StridedLayoutAttr::get(sourceMemRefType.getContext(),
