diff --git a/mlir/include/mlir/Transforms/Passes.h b/mlir/include/mlir/Transforms/Passes.h
index 83ccaf5d06f3..26492db968ed 100644
--- a/mlir/include/mlir/Transforms/Passes.h
+++ b/mlir/include/mlir/Transforms/Passes.h
@@ -51,7 +51,6 @@ class GreedyRewriteConfig;
 /// Creates an instance of the Canonicalizer pass, configured with default
 /// settings (which can be overridden by pass options on the command line).
 std::unique_ptr<Pass> createCanonicalizerPass();
-std::unique_ptr<Pass> createCanonicalizerPass(const CanonicalizerOptions &options);
 
 #if BSPUB_DAVINCI_BISHENGIR
 std::unique_ptr<Pass>
